packages/did-core-test-server/suites/did-spec/did-core-properties.js

```diff
@@ -0,0 +1,382 @@
+            expect(serviceValue.type.length > 0).toBe(true);
+          } else if(Array.isArray(serviceValue.type)) {
+            const types = serviceValue.type;
+            types.forEach(type => {
+              expect(typeof type === 'string').toBe(true);
+            });
+          } else {
+            throw new Error('Invalid value for `type` property.');
+          }
+        });
+      }
+  });
+
+  it('5.4 Services - The value of the serviceEndpoint property MUST be ' +
+    'a string, a map, or an ordered set composed of one or more strings ' +
+    'and/or maps.', async () => {
+      const {service} = didDocument;
+      if(service) {
+        service.forEach(serviceValue => {
+          const {serviceEndpoint} = serviceValue;
+          if(typeof serviceEndpoint === 'string') {
+            expect(serviceEndpoint.length > 0).toBe(true);
+          } else if(typeof serviceEndpoint === 'object') {
+            expect(Object.entries(serviceEndpoint).length > 0).toBe(true);
+          } else if(Array.isArray(serviceEndpoint)) {
+            serviceEndpoint.forEach(endpointValue => {
+              if(typeof endpointValue === 'string') {
+                expect(endpointValue.length > 0).toBe(true);
+              } else if(typeof endpointValue === 'object') {
+                expect(Object.entries(endpointValue).length > 0).toBe(true);
+              }
+            });
+          } else {
+            throw new Error(
+              'Unknown value for serviceEndpoint: ' + serviceEndpoint);
+          }
+        });
+      }
+  });
+
+  it('5.4 Services - All [serviceEndpoint] string values MUST be valid ' +
+    'URIs conforming to [RFC3986] and normalized according to the ' +
+    'Normalization and Comparison rules in RFC3986 and to any normalization ' +
+    'rules in its applicable URI scheme specification.', async () => {
+      const {service} = didDocument;
+      if(service) {
+        service.forEach(serviceValue => {
+          const {serviceEndpoint} = serviceValue;
+          if(typeof serviceEndpoint === 'string') {
+            expect(isValidURI(serviceEndpoint)).toBe(true);
+          }
+        });
+      }
+  });
+}
+
+const getAllVerificationMethods = (didDocument) => {
+  const verificationMethods = [];
+  const verificationRelationships = ['verificationMethod', 'authentication',
+    'assertionMethod', 'keyAgreement', 'capabilityInvocation',
+    'capabilityDelegation'];
+  verificationRelationships.forEach(name => {
+    didDocument[name].forEach(vr => {
+      if(typeof vr === 'object') {
+        verificationMethods.push(vr);
+      }
+    });
+  })
+
+  return verificationMethods;
+};
+
+const didCorePropertiesTests = (suiteConfig) => {
+  describe('did-core-properties', () => {
+    suiteConfig.dids.forEach((did) => {
+      describe(did, () => {
+        for(const [mediaType, resolutionResult] of Object.entries(suiteConfig[did])) {
+          if(jsonMediaTypes.includes(mediaType)) {
```

Comment:
What does this do? It feels strange that testing of Core Properties somehow depends on media types?

Reply:
We discussed this issue on the working group call, its an artifact of the fact that we are:

1. using json fixtures for the ADM
2. testing serialization of the ADM in JSON (did+ld+json and did+json)

I spent a good amount of time thinking about how to make this better without increasing the burden on test implementers, and this is what I came up with.